aten/src/ATen/native/cuda/MultiMarginLoss.cu

```diff
@@ -0,0 +1,388 @@
+        C10_CUDA_KERNEL_LAUNCH_CHECK();
+      } else if (p == 2) {
+        cunn_MultiMarginCriterion_updateOutput_kernel<2> <<<blocks, threads, 0, stream>>>(
+            out.data_ptr<scalar_t>(),
+            input.data_ptr<scalar_t>(),
+            target.data_ptr<int64_t>(),
+            weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+            1,
+            input.dim() < 1 ? input.numel() : input.sizes()[0],
+            reduction == at::Reduction::Mean,
+            margin);
+        C10_CUDA_KERNEL_LAUNCH_CHECK();
+      }
+    } else if (input.dim() == 2) {
+      auto in_sizes = input.sizes();
+      int nframe = in_sizes[0];
+      // allow zero-dim target for 2D input.
+      TORCH_CHECK(in_sizes[1] != 0 && target.dim() <= 1 && target.numel() == nframe,
+                "inconsistent target size");
+      dim3 blocks(nframe);
+      dim3 threads(MULTIMARGIN_THREADS);
+
+      if (reduction == at::Reduction::None) {
+        if (p == 1) {
+          cunn_MultiMarginCriterion_updateOutput_kernel<1> <<<blocks, threads, 0, stream>>>(
+              out.data_ptr<scalar_t>(),
+              input.data_ptr<scalar_t>(),
+              target.data_ptr<int64_t>(),
+              weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+              nframe, in_sizes[1],
+              false,
+              margin);
+          C10_CUDA_KERNEL_LAUNCH_CHECK();
+        } else if (p == 2) {
+          cunn_MultiMarginCriterion_updateOutput_kernel<2> <<<blocks, threads, 0, stream>>>(
+              out.data_ptr<scalar_t>(),
+              input.data_ptr<scalar_t>(),
+              target.data_ptr<int64_t>(),
+              weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+              nframe, in_sizes[1],
+              false,
+              margin);
+          C10_CUDA_KERNEL_LAUNCH_CHECK();
+        }
+      } else {
+        auto tmp_output = at::empty({nframe}, input.options());
+        if (p == 1) {
+          cunn_MultiMarginCriterion_updateOutput_kernel<1> <<<blocks, threads, 0, stream>>>(
+              tmp_output.data_ptr<scalar_t>(),
+              input.data_ptr<scalar_t>(),
+              target.data_ptr<int64_t>(),
+              weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+              nframe, in_sizes[1],
+              reduction == Reduction::Mean,
+              margin);
+          C10_CUDA_KERNEL_LAUNCH_CHECK();
+        } else if (p == 2) {
+          cunn_MultiMarginCriterion_updateOutput_kernel<2> <<<blocks, threads, 0, stream>>>(
+              tmp_output.data_ptr<scalar_t>(),
+              input.data_ptr<scalar_t>(),
+              target.data_ptr<int64_t>(),
+              weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+              nframe, in_sizes[1],
+              reduction == Reduction::Mean,
+              margin);
+          C10_CUDA_KERNEL_LAUNCH_CHECK();
+        }
+        at::sum_out(out, tmp_output, /*dims=*/IntArrayRef{});
+      }
+    } else {
```

Comment:
can this ever happen? input.ndim() <= 2 is checked in the multi_margin_loss_shape_check.

Reply:
Yeah, I think you're right. I've changed it to an internal assert in the 2d case.